### PR TITLE
py-cryptography: Update to 3.4.7

### PR DESCRIPTION
**Pull request by essandess**
###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS x.y
Xcode x.y

###### Verification 
Have you

- [ ] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [ ] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [ ] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by stromnov** [comment]
cryptography 3.x is backward incompatible with macOS 10.9 and below. So it's not possible to simply bump the library version without affecting old macOSes.

https://lists.macports.org/pipermail/macports-dev/2020-October/042505.html

**Comment by essandess** [comment]
Let’s get on the major update for the great majority of deployments and backfill the old edge cases later. Perhaps create a port `py-cryptography2`.

**Comment by stromnov** [comment]
As far as I see, there will be too many edge cases. Every port, which depends on py-cryptography, will be an edge case, requiring substitution py-cryptography with py-cryptography2 (or -legacy) if OS version <= 10.9.

**Comment by essandess** [comment]
This port is out-of-date with several known insecurities: https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst

May we please update ASAP?

If anyone wants to install insecure, out-of-date versions they may choose to do so, but MacPorts shouldn’t be distributing cryptography code with unpatched CVEs.

**Comment by kencu** [comment]
builds are failing due to:
```
2021-04-28T22:31:15.2833650Z ModuleNotFoundError: No module named 'setuptools_rust'
2021-04-28T22:31:15.2967450Z Command failed:  cd "/opt/local/var/macports/build/_Users_runner_work_macports-ports_macports-ports_ports_python_py-cryptography/py38-cryptography/work/cryptography-3.4.7" && /opt/local/Library/Frameworks/Python.framework/Versions/3.8/bin/python3.8 setup.py --no-user-cfg build -j3 
```

First I've seen of "setuptools_rust" -- oh what fun, if python ports now need rust to build :>

**Comment by stromnov** [comment]
> You will also need to have Rust installed and available, which can be obtained from Homebrew, MacPorts, or directly from the Rust website.

https://cryptography.io/en/latest/installation/#building-cryptography-on-macos

**Comment by herbygillot** [comment]
> First I've seen of "setuptools_rust" -- oh what fun, if python ports now need rust to build :>

Yep, this has caused plenty of Internet entertainment. You can probably start with  https://github.com/pyca/cryptography/issues/5771

**Comment by essandess** [comment]
Looks like we should also remove Python 35 (and 36 too), but there are dependencies to chase down.

https://github.com/pyca/cryptography/blob/33bc4fd2394b0168508cff1e8a0c6645528d47de/setup.py#L90

I suggest we leave these for another PR and expunge the CVEs.

```bash
port echo rdepends:py35-cryptography
py35-ansible                    
py35-ansible-base               
py35-comet                      
py35-fido2                      
py35-libcloud                   
py35-oauthlib                   
py35-openssl                    
py35-paramiko                   
py35-requests-oauthlib          
py35-service_identity           
py35-trustme                    
py35-tweepy                     
py35-twilio
```

**Comment by reneeotten** [comment]
this newer version is Python 3.6+ so it will need to take care of PY27 ports as well. There needs to be a plan in place for how to either remove all the (recursive) dependencies of these ports or pin them to an earlier version. I think that should be part of this PR and not postponed to a later point; merging this will likely break things and the required clean-ups might not happen anytime soon if not done now. Additionally, there is the point of breaking for 10.9 and below that @stromnov mentions - what's the opinion on that?

**Comment by herbygillot** [comment]
Does `setuptools-rust` still need to be in `depends_lib-append`, or can it live in `depends_build-append`?

**Comment by essandess** [comment]
> Does `setuptools-rust` still need to be in `depends_lib-append`, or can it live in `depends_build-append`?

I put it in `depends_build-append`, added `cargo_fetch` stuff, and added tests. The tests all run in Python 39.

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:47]: why is `setuptools` not only a build dependency; and were do `py-six` and `py-pip` come from, I don't see them listedupstream, or...?

**Review by reneeotten — CHANGES_REQUESTED** [review]
before a possible merge or this PR, this will need a solution for unsupported Python versions and possibly older systems and, importantly, agreement from the current maintainer.

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:44]: also since version 2.8 it seems that `asn1crypto` is no longer a [dependency](https://github.com/pyca/cryptography/blob/b2f9a0e1a546c3b10a6e96cd5e8423a7b04970c7/CHANGELOG.rst#28---2019-10-16).

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:39]: this only has subports for py37/38/39

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:7]: are you actually using this PG?

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:39]: That’s what upstream supports, ignoring EOL 36:

https://github.com/PyO3/setuptools-rust/blob/61b0603ba8cecaff34331904de8f2a0d035152ca/setup.cfg#L31

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:7]: “`cargo.crates`” avoids rust fetches after the fetch phase.

**Comment by essandess** [comment]
> before a possible merge or this PR, this will need a solution for unsupported Python versions and possibly older systems and, importantly, agreement from the current maintainer.

Agreed that EOL Python dependencies should be removed. The major affected port looks to be mailman’s, but this port is out-of-date and must be updated to `mailman3`. I’ll defer to other whether to continue support for `mailman2` and how to accomplish that.

Older systems are already explicitly not supported by upstream—this happened four years ago and isn’t up to us: https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#19---2017-05-29

The current version that MacPorts distributes has several known CVEs. We must update to the latest version as soon as possible.

File: python/py-cryptography/Portfile
Comment (reneeotten) [reply python/py-cryptography/Portfile:39]: sure, but that means that you will need to remove the PY27/PY35 subports from `py-cryptography` (which doesn't support lower versions either) and add PY36 to `py-setuptools-rust`

Comment (reneeotten) [reply python/py-cryptography/Portfile:39]: and Python 3.6 is not EOL [yet](https://devguide.python.org/#status-of-python-branches)

File: python/py-cryptography/Portfile
Comment (reneeotten) [reply python/py-cryptography/Portfile:7]: sorry for not being completely clear, I was referring to the `compiler_blacklist_versions` PG

**Comment by kencu** [comment]
If this port is merged without any backup in place, and all systems that can't support rust are dropped, just what kind of rdeps wreckage is expected here?

One of MacPorts unique strengths is it is still useful to all those people out there who have older systems they are not ready to bin yet. It would be a shame to cut them all off because of what amounts to one guy's rampage about C/C++ being evil languages that need to be dropped in favour of rust.

So taking a bit of time to come up with a coherent plan is a good idea. There is nothing urgent about this.

**Comment by kencu** [comment]
IMHO, of course. I clearly don't speak for MacPorts, nor have any particular sway in the mix.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: I put `setuptools` in `depends_build`. I put `pip` in there because `port -vst install` failed without it and complained explicitly about not finding `pip`.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: Thanks fixed.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:37]: Right, not dead yet. I kept it in the latest push.

**Comment by essandess** [comment]
> If this port is merged without any backup in place

I just added legacy support. Old dependencies just need to point at `py*-cryptography2`.

**Comment by kencu** [comment]
> Not serving up CVEs takes precedence many times over the tradeoff of supporting old systems that are explicitly not supported upstream. This isn’t MacPorts pulling the plug, upstream pulled it years ago. It’s out of our hands.

No debate about that.

But leaving behind the last non-rust version for those of us who don't work in major multinational bank or at the NSA, and who just want our kids to be able to play some games on the old computer or use it to read some books, is not perhaps a terribly big ask?

File: python/py-cryptography/Portfile
Comment (reneeotten) [reply python/py-cryptography/Portfile]: sure I was suggesting to put `py-setuptools` in `depends_build` instead of `depends_lib` where you had it initially since it shouldn't be required at runtime. I still don't see where `pip` comes in, that shouldn't be needed I would think; please show the output of the build failure.

File: python/py-cryptography/Portfile
Comment (reneeotten) [reply python/py-cryptography/Portfile]: I still don't see where the `compiler_blacklist_versions` PG is used in that subport either; I might be missing it, so please point me to it

**Comment by reneeotten** [comment]
> > If this port is merged without any backup in place
> 
> I just added legacy support. Old dependencies just need to point at `py*-cryptography2`.

please make a separate port for `py-cryptography2`; just putting everything into one file doesn't help readability/maintainability.

I agree with @kencu that having a solid plan first before merging this makes sense. This port has been outdated for years without anyone being bothered about it - so a few more days or a week shouldn't be an issue. 

So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.

**Comment by essandess** [comment]
> So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.

Please see #10873.

**Comment by reneeotten** [comment]
> > So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.
> 
> Please see #10873.

Thanks @essandess - this will take care of Python 2.7 ports to use the latest supported version of `py-cryptography`. It doesn't deal yet with the fact that older systems that cannot install `rust` will have no way of installing any of the Python 3.x subports anymore - correct? 

I think that's the other remaining point that also @kencu and @stromnov  pointed out. To be able to build these Python ports on older systems, I think one would need to do the dependency replacement when `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`, where `XY` is the latest version where `rust` builds.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: I'm honestly not sure about that—seems like it was necessary with trace mode. I'll delete it and hope for the best.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: Removed thanks.

**Comment by essandess** [comment]
> > > So what's the plan forward now: are you suggesting to change all dependents of this port to `py-cryptography2` for systems that cannot build `rust`? If so, please do that in this PR.
> > 
> > 
> > Please see #10873.
> 
> Thanks @essandess - this will take care of Python 2.7 ports to use the latest supported version of `py-cryptography`. It doesn't deal yet with the fact that older systems that cannot install `rust` will have no way of installing any of the Python 3.x subports anymore - correct?
> 
> I think that's the other remaining point that also @kencu and @stromnov pointed out. To be able to build these Python ports on older systems, I think one would need to do the dependency replacement when `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`, where `XY` is the latest version where `rust` builds.

My request is that further legacy details be addressed later because it involves a lot of MacPorts system details with which I am unfamiliar, takes a lot of time to rebase this many commits, and further delays elliminating the version 2 CVE from the MacPorts distribution.

**Comment by essandess** [comment]
Re: `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`

Upstream [says](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#30---2020-07-20) <blockquote>**BACKWARDS INCOMPATIBLE:** Dropped support for macOS 10.9, macOS users must upgrade to 10.10 or newer.</blockquote>

What's the `${os.major}` for macOS 10? If you all agree on the exact logic, I volunteer to rebase the commits in #10873 to use this logic.

**Comment by reneeotten** [comment]
> Update: Please see all the commits in #10873. This is essentially represented in all the relevant dependencies:
> 
> ```
>     if {${python.version} eq 27} {
>         depends_lib-replace \
>                     port:py${python.version}-cryptography \
>                     port:py${python.version}-cryptography2
>     } 
> ```

yes, that will take care of using the previous version for PY27; but what happens if I would want to install the PY39 subport of one of these ports on a system that doesn't support `rust`?

**Comment by essandess** [comment]
> Re: `if {{${python.version} eq 27} or {${os.platform} eq "darwin" && ${os.major} < XY}}`
> 
> Upstream [says](https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst#30---2020-07-20)
> 
> > **BACKWARDS INCOMPATIBLE:** Dropped support for macOS 10.9, macOS users must upgrade to 10.10 or newer.
> 
> What's the `${os.major}` for macOS 10? If you all agree on the exact logic, I volunteer to rebase the commits in #10873 to use this logic.

From [here](https://trac.macports.org/wiki/PortfileRecipes) it's `14`. I will use this logic in #10873 unless someone requests an alternate in the next 1–2 days:
```
    if {${python.version} eq 27 || {${os.platform} eq "darwin" && ${os.major} < 14}} {
        depends_lib-replace \
                    port:py${python.version}-cryptography \
                    port:py${python.version}-cryptography2
    } 
```

**Comment by reneeotten** [comment]
@essandes - great, that would cover everything I think. From the [buildbots](https://ports.macports.org/port/rust/summary) it looks like `rust` did build on 10.9 so perhaps the latest version of that port can still be used there? Of course, I am not sure if there is something specifically in the `py-cryptography` port that would require 10.10+.

**Comment by essandess** [comment]
The latest [force-push](https://github.com/macports/macports-ports/compare/4c5a2dc81bd8e7d944a663db03ead00ab3502527..4a111a52bcfed9211c5a11b25c7db175d0ff2504) implements the suggestions above and in #10873.

**Comment by essandess** [comment]
@reneeotten These CI failures are why I believe that a pip build dependency is required:
```

Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/setuptools/installer.py", line 75, in fetch_build_egg
    subprocess.check_call(cmd)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_runner_work_1_s_ports_python_py-cryptography/py39-cryptography/work/.tmp/tmpeb_zl60_', '--quiet', '--index-url', "''", '--find-links', "''", 'toml>=0.9.0']' returned non-zero exit status 1.
```

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: These trace mode and CI failures are why I believe that a pip build dependency is required:
```
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/setuptools/installer.py", line 75, in fetch_build_egg
    subprocess.check_call(cmd)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_runner_work_1_s_ports_python_py-cryptography/py39-cryptography/work/.tmp/tmpeb_zl60_', '--quiet', '--index-url', "''", '--find-links', "''", 'toml>=0.9.0']' returned non-zero exit status 1.
```

**Comment by reneeotten** [comment]
> @reneeotten These CI failures are why I believe that a pip build dependency is required:
> 
> ```
> Traceback (most recent call last):
>   File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/setuptools/installer.py", line 75, in fetch_build_egg
>     subprocess.check_call(cmd)
>   File "/opt/local/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 373, in check_call
>     raise CalledProcessError(retcode, cmd)
> subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.9/bin/python3.9', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_runner_work_1_s_ports_python_py-cryptography/py39-cryptography/work/.tmp/tmpeb_zl60_', '--quiet', '--index-url', "''", '--find-links', "''", 'toml>=0.9.0']' returned non-zero exit status 1.
> ```

That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).

File: python/py-pytest-subtests/Portfile
Comment (reneeotten) [inline python/py-pytest-subtests/Portfile:16]: this also needs a PY36 subport since it's needed for `py36-cryptography`, see the `lint` error: " Error: Unknown dependency: py36-pytest-subtests"

File: python/py-pytest-subtests/Portfile
Comment (reneeotten) [inline python/py-pytest-subtests/Portfile:14]: please take this from PyPI instead to avoid the `fetch.type git`

Comment (essandess) [reply python/py-pytest-subtests/Portfile:14]: Okay, but pretty sure that will break things.

Would you please step in and survey the breakage and offer a fix? There will be some complaint from the build command that it's not a valid git repo.

I could never figure that one out, so I just used a git clone to make this issue go away.

**Comment by essandess** [comment]
> That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).

Thanks. There now a few (one) ports that use `setuptools-rust` that have this issue. Pinning this to fix later: https://github.com/macports/macports-ports/blob/master/python/py-tokenizers/Portfile

**Comment by essandess** [comment]
> That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).

Shouldn't `setuptools-rust` simply add `toml` as a build dependency? Any objections to just adding that fix to this PR?

File: python/py-pytest-subtests/Portfile
Comment (essandess) [reply python/py-pytest-subtests/Portfile]: Thanks done.

File: python/py-pytest-subtests/Portfile
Comment (essandess) [reply python/py-pytest-subtests/Portfile]: Done. I predict a CI failure. Would you please take a look when it comes?

**Comment by reneeotten** [comment]
> > That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).
> 
> Shouldn't `setuptools-rust` simply add `toml` as a build dependency? Any objections to just adding that fix to this PR?

It already has, but that doesn't help building in tracemode because `py-toml` isn't specified there or on the buildbots because the ports are being uninstalled in between installations.

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:39]: this dependency is only needed for version 2.9.2 so it should be moved down into the "legacy" block

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:68]: this dependency isn't needed anymore since version 2.8.0 as I pointed out earlier I think

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:52]: why are you switching now to PyPI for this part and not use GitHub as well?

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:29]: this isn't very clear for future reference... the point is that the minimum version here should be the OS version where `rust` still builds - correct?

**Comment by essandess** [comment]
> > > That message is telling you that it is missing `toml` and tries to install that with `pip`, which fails because `pip` isn't there; the solution is to add `py-toml` to the build dependencies instead. Upstream has not listed this explicitly as a dependency, however it's pulled in indirectly through `setuptools-rust`; building in tracemode or on the buildbots without specifying it in this port will make the builds fail. (There are a few other remaining points that I'll review later).
> > 
> > 
> > Shouldn't `setuptools-rust` simply add `toml` as a build dependency? Any objections to just adding that fix to this PR?
> 
> It already has, but that doesn't help building in tracemode because `py-toml` isn't specified there or on the buildbots because the ports are being uninstalled in between installations.

I see that @cjones051073 added it as a build dependency. We need it as a run dependency. That means we should add it as a lib dependency, I believe. I will rebase everything to include @cjones051073's commit and add this.

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:84]: add `py-toml` here

**Review by reneeotten — COMMENTED** [review]
hopefully the last round

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:84]: I think toml should be a lib dependency of setuptools-rust, which will fix this issue here and other ports that use it.

File: python/py-setuptools-rust/Portfile
Comment (reneeotten) [inline python/py-setuptools-rust/Portfile:34]: this shouldn't be added here since it is not needed at runtime as far as I can tell (and if you would want to add it also as runtime dependency, it should be listed once under `depends_lib` and then remove from `depends_build`)

File: python/py-cryptography/Portfile
Comment (reneeotten) [reply python/py-cryptography/Portfile:84]: if indeed *all* ports that use `setuptools-rust` need also `toml`, then yes that could be a solution. 

It seems a bit similar though to the `setuptools_scm` situation, where sometimes packages specify the dependency as `setuptools_scm[toml]` indicating that `toml` is also needed. The question is then where to add the `toml` dependency; I would say explicitly in the packages that need it, and not in `setuptools_scm` (or `setuptools-rust`), but I can see both sides and don't really care.

File: python/py-pytest-subtests/Portfile
Comment (reneeotten) [reply python/py-pytest-subtests/Portfile]: you only removed `fetch.type git`, that will likely not work indeed. I suggested to take the file from PyPI, which *should* work correctly, but for that to happen you'll need to remove the `github` PG and associated stuff as well.

Comment (essandess) [reply python/py-pytest-subtests/Portfile]: I tried switching to PyPI before but it throws the same error. I'll amend to demonstrate…

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: Thanks done.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: Thanks done.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: Clarified thanks.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: I included a commit to setuptools-rust because I recall this issue in the tokenizers port.

File: python/py-setuptools-rust/Portfile
Comment (essandess) [reply python/py-setuptools-rust/Portfile]: Moved to depends_lib

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile]: Now using git for everything, thanks.

File: python/py-pytest-subtests/Portfile
Comment (essandess) [reply python/py-pytest-subtests/Portfile]: To the best of my knowledge, this is the last remaining major issue. I'm expecting a CI build failure using PyPI now…

Comment (essandess) [reply python/py-pytest-subtests/Portfile]: Looks like I was wrong about the PyPI source, and this all builds fine.

**Comment by essandess** [comment]
@reneeotten I believe this is ready to merge. May we all please declare success?

File: python/py-pytest-subtests/Portfile
Comment (reneeotten) [inline python/py-pytest-subtests/Portfile:21]: this is missing the `homepage` field as pointed out by `port lint`

File: python/py-pytest-subtests/Portfile
Comment (reneeotten) [inline python/py-pytest-subtests/Portfile:40]: if there are not tests on PyPI (which is fine), just remove this block instead of commenting it out

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:74]: this comment is not correct anymore; it says "install pyXY-cryptography for an up-to-date version"; but a user will get this message when exactly doing that when she/he is installing PY27 or on a system that doesn't support `rust`. So saying that it has a CVE in the package is fine, but in the end of the day there isn't anything the user (or we) can do about it.

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:60]: what's the purpose of this block... can this not just be `${os.platform} eq "darwin" && ${os.major} < 11`?

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:45]: please align this

File: python/py-cryptography/Portfile
Comment (reneeotten) [inline python/py-cryptography/Portfile:29]: so `rust` still builds on 10.9 according to [this](https://ports.macports.org/port/rust/summary), so please set this to `13` for now and we'll see what the buildbots make of it.

**Review by reneeotten — CHANGES_REQUESTED** [review]
@essandess we can declare almost success I think!

File: python/py-pytest-subtests/Portfile
Comment (essandess) [reply python/py-pytest-subtests/Portfile:40]: Thanks done.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:74]: Thanks done.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:45]: Thanks done.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:29]: Thanks done.

File: python/py-cryptography/Portfile
Comment (essandess) [reply python/py-cryptography/Portfile:60]: Legacy code? I changed it to what you suggest.

**Pull request merged**
